### ISSUE TITLE
__array_interface__ does not maintain persistance of memory location for numpy arrays

### ISSUE THREAD

**Issue by lhfriedman** [numpy/numpy#16713]
Given that `__array_interface__` is primarily to enable use of things such as np.asarray([USER_CLASS_INSTANCE]), I am guessing the following code is unintended behavior related to timing of garbage collection. I ran across it while trying to enable 
`np.asarray([A_DATACLASS_OBJECT_WITH_FLOAT_ARRAY_MEMBERS])`

### Reproducing code example:

```python
import numpy as np

class MyClass:
    """
    Precursor to numpy array with first element, a_0, and second element, a_1.
    """

    def __init__(self, n, retain_array):
        self.a_0 = np.arange(n)
        self.a_1 = np.arange(n)
        self.retain_array=retain_array

    @property
    def __array_interface__(self):
        arr_0 = np.array([self.a_0, self.a_1])
        if self.retain_array:
            self.temp = arr_0
        return arr_0.__array_interface__

    def __str__(self):
        out = f"""
        MyClass(a_0={self.a_0[:5].tolist()}...{self.a_0[-5:].tolist()},
                a_1={self.a_1[:5].tolist()}...{self.a_0[-5:].tolist()})
        """
        return out

print('\nnumpy version: ', np.version.full_version)

print(80*'=')
print('No problem with small arrays.')
mc_0 = MyClass(10, False)
print('\nmc_0:\n', mc_0)
mc_1 = np.asarray(mc_0)
print('\nmc_1[:, :5]:\n', mc_1)

print(80*'=')
print('Large arrays have problems')
mc_0 = MyClass(1000, False)
print('\nmc_0:\n', mc_0)
npa_0 = np.asarray(mc_0)
print('\nnpa0:\n',
      f"""
      [{npa_0[0, :5].tolist()}...{npa_0[0, -5:].tolist()},
       {npa_0[1, :5].tolist()}...{npa_0[1, -5:].tolist()}]
      """
      )

print(80*'=')
print(
      """Problem appears to come from peristance of array. With
      retain_array=True, no problem.
      """)
mc_0 = MyClass(1000, True)
print('\nmc_0:\n', mc_0)
npa_0 = np.asarray(mc_0)
print('\nnpa0:\n',
      f"""
      [{npa_0[0, :5].tolist()}...{npa_0[0, -5:].tolist()},
       {npa_0[1, :5].tolist()}...{npa_0[1, -5:].tolist()}]
      """
      )
```

### Output
```

numpy version:  1.18.1
================================================================================
No problem with small arrays.

mc_0:
 
        MyClass(a_0=[0, 1, 2, 3, 4]...[5, 6, 7, 8, 9],
                a_1=[0, 1, 2, 3, 4]...[5, 6, 7, 8, 9])
        

mc_1[:, :5]:
 [[0 1 2 3 4 5 6 7 8 9]
 [0 1 2 3 4 5 6 7 8 9]]
================================================================================
Large arrays have problems

mc_0:
 
        MyClass(a_0=[0, 1, 2, 3, 4]...[995, 996, 997, 998, 999],
                a_1=[0, 1, 2, 3, 4]...[995, 996, 997, 998, 999])
        

npa0:
 
      [[-172283504, 528, -172825152, 528, 4]...[995, 996, 997, 998, 999],
       [0, 1, 2, 3, 4]...[995, 996, 997, 998, 999]]
      
================================================================================
Problem appears to come from peristance of array. With
      retain_array=True, no problem.
      

mc_0:
 
        MyClass(a_0=[0, 1, 2, 3, 4]...[995, 996, 997, 998, 999],
                a_1=[0, 1, 2, 3, 4]...[995, 996, 997, 998, 999])
        

npa0:
 
      [[0, 1, 2, 3, 4]...[995, 996, 997, 998, 999],
       [0, 1, 2, 3, 4]...[995, 996, 997, 998, 999]]
      
```

### Numpy/Python version information:

1.18.1 3.7.7 (default, May  6 2020, 11:45:54) [MSC v.1916 64 bit (AMD64)]

**Comment by eric-wieser** [comment]
Is using `__array__` instead an option for you?

**Comment by eric-wieser** [comment]
Your diagnosis is correct though, `arr_0.__array_interface__` either doesn't contain the information to keep arr_0 alive, or that information is being discarded.

**Comment by seberg** [comment]
`__array_interface__` is only supposed to be used by object which own the memory themselves (of course that part could be lazy).  In all other cases you should indeed use `__array__`.

**Comment by lhfriedman** [comment]
Thank you for your very fast reply, and also, thank you for code that has been immensely helpful for a number of research projects.

The `__array__` method works, but I didn't know it was an option. When I searched on it, I found (https://numpy.org/devdocs/reference/arrays.classes.html?highlight=__array__#numpy.class.__array__). It might just be me, but the documentation appears to encourage use of `__array_interface__` when `__array__` is more broadly applicable. (https://numpy.org/devdocs/reference/arrays.interface.html). 

I humbly suggest that the documentation for `__array__` be expanded, Just something like.  

> Adding an  `__array__` method to a class definition allows for conversion of class instances to numpy.ndarray in a manner similar to native Python lists and tuples.  `__array__` should take only `self` as an argument and return the desired numpy.ndarray instance.

 and that the documentation for `__array_interface__` suggest use of `__array__` if possible.  


The following class definition works fine, but where is this indicated in the documentation?

Please go ahead and close the issue if you think this settles it, but I guess I am interested in hearing what you think of my suggested documentation changes, Yay or nay?

```
class MyClass:
    """
    Precursor to numpy array with first element, a_0, and second element, a_1.
    """

    def __init__(self, n, retain_array):
        self.a_0 = np.arange(n)
        self.a_1 = np.arange(n)
        self.retain_array=retain_array

    # @property
    # def __array_interface__(self):
    #     arr_0 = np.array([self.a_0, self.a_1])
    #     if self.retain_array:
    #         self.temp = arr_0
    #     return arr_0.__array_interface__


    def __array__(self):
        return np.array([self.a_0, self.a_1])

    def __str__(self):
        out = f"""
        MyClass(a_0={self.a_0[:5].tolist()}...{self.a_0[-5:].tolist()},
                a_1={self.a_1[:5].tolist()}...{self.a_0[-5:].tolist()})
        """
        return out
```

**Comment by mattip** [comment]
I think the point is that `__array_interface__` implements a protocol to expose internal structures from a Python object, where `__array__` is used to create a new object with an "array-like" interface (I think, it is not really clear what `__array__` does). They each have their use. The documentation for [`__array__`](https://numpy.org/devdocs/reference/arrays.classes.html?highlight=__array__#numpy.class.__array__) is not very useful.

**Comment by rossbar** [comment]
> The documentation for __array__ is not very useful.

Related issue: #16330

**Comment by lhfriedman** [comment]
Seems to be addressed in related issue #16330
